internal/pkg/skuba/addons/gangway_test.go

```diff
@@ -0,0 +1,146 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package addons
+
+import (
+	"regexp"
+	"testing"
+
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubernetes"
+	"github.com/SUSE/skuba/internal/pkg/skuba/skuba"
+	img "github.com/SUSE/skuba/pkg/skuba"
+)
+
+func TestGetGangwayImage(t *testing.T) {
+	tests := []struct {
+		name     string
+		imageTag string
+		want     string
+	}{
+		{
+			name:     "get gangway init image without revision",
+			imageTag: "3.1.0",
+			want:     img.ImageRepository + "/gangway:3.1.0",
+		},
+		{
+			name:     "get gangway init image with revision",
+			imageTag: "3.1.0-rev2",
+			want:     img.ImageRepository + "/gangway:3.1.0-rev2",
+		},
+	}
+	for _, tt := range tests {
+		tt := tt // Parallel testing
+		t.Run(tt.name, func(t *testing.T) {
+			if got := GetGangwayImage(tt.imageTag); got != tt.want {
+				t.Errorf("GetGangwayImage() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_renderContext_GangwayImage(t *testing.T) {
+	type test struct {
+		name          string
+		renderContext renderContext
+		want          string
+	}
+	for _, ver := range kubernetes.AvailableVersions() {
+		tt := test{
+			name: "render gangway image url when cluster version is " + ver.String(),
+			renderContext: renderContext{
+				config: AddonConfiguration{
+					ClusterVersion: ver,
+					ControlPlane:   "",
+					ClusterName:    "",
+				},
+			},
+			want: img.ImageRepository + "/gangway:([[:digit:]]{1,}.){2}[[:digit:]]{1,}(-rev[:digit:]{1,})?",
+		}
+		t.Run(tt.name, func(t *testing.T) {
+			got := tt.renderContext.GangwayImage()
+			matched, err := regexp.Match(tt.want, []byte(got))
+			if err != nil {
+				t.Error(err)
+				return
+			}
+			if !matched {
+				t.Errorf("renderContext.GangwayImage() = %v, want %v", got, tt.want)
+				return
+			}
+		})
+	}
+}
+
+func Test_gangwayCallbacks_beforeApply(t *testing.T) {
+	type test struct {
+		name               string
+		addonConfiguration AddonConfiguration
+		skubaConfiguration *skuba.SkubaConfiguration
+		wantErr            bool
+	}
+	for _, ver := range kubernetes.AvailableVersions() {
+		tt := test{
+			name: "test gangway callbacks beforeApply when cluster version is " + ver.String(),
+			addonConfiguration: AddonConfiguration{
+				ClusterVersion: ver,
+				ControlPlane:   "",
+				ClusterName:    "",
+			},
+			skubaConfiguration: &skuba.SkubaConfiguration{
+				AddonsVersion: kubernetes.AddonsVersion{},
+			},
+			wantErr: true,
+		}
+		t.Run(tt.name, func(t *testing.T) {
+			g := gangwayCallbacks{}
+			if err := g.beforeApply(tt.addonConfiguration, tt.skubaConfiguration); (err != nil) != tt.wantErr {
+				t.Errorf("gangwayCallbacks.beforeApply() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func Test_gangwayCallbacks_afterApply(t *testing.T) {
```

Comment:
Right now the `afterApply` https://github.com/SUSE/skuba/blob/master/internal/pkg/skuba/addons/gangway.go#L82-L84 is return nil, I don't know what this test case what to test?

Reply:
removed